### PR TITLE
[controller-manager] Switch main `Project` controller to `controller-runtime`

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area dev-productivity scalability
/kind enhancement

**What this PR does / why we need it**:

- Refactor the main `Project` controller to `controller-runtime`.
- Cleanup technical debt in the `Project` controller
- upgrade controller-runtime to `v0.12.3`
- add an e2e test for project extension roles
- small changes to how ginkgo reports in unit and integration tests

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/4251

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by timebertt** [comment]
I stress-tested the new integration test a bit to hopefully not introduce the next flaky test:
```
$ stress -ignore "(unable to grab random port)" -p 32 ./test/integration/controllermanager/project/project/project.test -ginkgo.progress
...
3m45s: 96 runs so far, 0 failures
```

**Comment by timebertt** [comment]
I dropped 8a13cd78217b1238ae95ae27330e2c447489ef54 again, because it prevents `go test` from caching test results.

File: test/integration/controllermanager/project/project/project_test.go [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+		Expect(testClient.Create(ctx, shoot)).To(Succeed())
+		log.Info("Created Shoot for test", "shoot", client.ObjectKeyFromObject(shoot))
+
+		DeferCleanup(func() {
+			By("Cleanup Shoot")
+			Expect(client.IgnoreNotFound(testClient.Delete(ctx, shoot))).To(Succeed())
+			Eventually(func() error {
+				return testClient.Get(ctx, client.ObjectKeyFromObject(shoot), shoot)
+			}).Should(BeNotFoundError())
+		})
```
Comment (acumino) [inline test/integration/controllermanager/project/project/project_test.go:176]: this is not required shoot is getting deleted anyway in lines 189-194.

File: pkg/controllermanager/controller/project/project/reconciler.go [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+	if len(project.Status.Phase) == 0 {
+		if err := patchProjectPhase(ctx, gardenClient, project, gardencorev1beta1.ProjectPending); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	ownerReference := metav1.NewControllerRef(project, gardencorev1beta1.SchemeGroupVersion.WithKind("Project"))
+
+	// We reconcile the namespace for the project: If the .spec.namespace is set then we try to claim it, if it is not
+	// set then we create a new namespace with a random hash value.
```
Comment (acumino) [inline pkg/controllermanager/controller/project/project/reconciler.go]:
```suggestion
	// We reconcile the namespace for the project: If the .spec.namespace is set then we try to claim it, if it is not
	// set then we create a new namespace with a random hash value.
```
With this PR it is not random anymore and depends on project UID.

File: test/integration/controllermanager/project/project/project_test.go (hunk 1 above)
Comment (timebertt) [reply test/integration/controllermanager/project/project/project_test.go:176]: My reasoning was to make sure the shoot is cleaned up even if the test fails before reaching line 189.
If we don't do it, the cleanup of the project will also fail. That might orphan resources when running against an existing gardener setup.

File: test/integration/controllermanager/project/project/project_test.go [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+						g.Expect(projectNamespace.OwnerReferences).To(BeEmpty())
+						g.Expect(projectNamespace.Labels).NotTo(Or(
+							HaveKey("project.gardener.cloud/name"),
+							HaveKey("gardener.cloud/role"),
+						))
+					}).Should(Succeed())
+				})
+			})
+
+			Context("namespace belongs to another project", func() {
```
Comment (rfranzke) [inline test/integration/controllermanager/project/project/project_test.go:330]: We could also add a test case for a namespace without the labels

File: pkg/controllermanager/controller/project/project/reconciler.go [hunk 4]
```diff
@@ -31,25 +44,49 @@ import (
...
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	if project.DeletionTimestamp != nil {
+		log.Info("Deleting project")
+		return r.delete(ctx, log, project, r.Client)
+	}
+
+	log.Info("Reconciling project")
+	return r.reconcile(ctx, log, project, r.Client)
```
Comment (rfranzke) [inline pkg/controllermanager/controller/project/project/reconciler.go]: No need to pass `r.Client`

**Review by rfranzke — COMMENTED** [review]
Very nice PR, loved reviewing it :) The tests are just awesome!

**Pull request merged**